activerecord/lib/active_record/railties/databases.rake

```diff
@@ -94,24 +94,43 @@ db_namespace = namespace :db do
 
   # IMPORTANT: This task won't dump the schema if ActiveRecord::Base.dump_schema_after_migration is set to false
   task :_dump do
-    if ActiveRecord::Base.dump_schema_after_migration
-      case ActiveRecord::Base.schema_format
-      when :ruby then db_namespace["schema:dump"].invoke
-      when :sql  then db_namespace["structure:dump"].invoke
-      else
-        raise "unknown schema format #{ActiveRecord::Base.schema_format}"
+    dump_all = ActiveRecord::Base.dump_schema_after_migration
+    database_configs = ActiveRecord::DatabaseConfigurations.new(databases).configs_for(env_name: Rails.env)
+
+    if database_configs.count == 1
+      if dump_all && database_configs.first.schema_dump
+        case ActiveRecord::Base.schema_format
+        when :ruby then db_namespace["schema:dump"].invoke
+        when :sql  then db_namespace["structure:dump"].invoke
+        else
+          raise "unknown schema format #{ActiveRecord::Base.schema_format}"
+        end
+      end
+    else
+      ActiveRecord::Tasks::DatabaseTasks.for_each(databases) do |name, db_config|
+        if dump_all && db_config.schema_dump
+          case ActiveRecord::Base.schema_format
+          when :ruby then db_namespace["schema:dump:#{name}"].invoke
+          when :sql  then db_namespace["structure:dump:#{name}"].invoke
+          else
+            raise "unknown schema format #{ActiveRecord::Base.schema_format}"
+          end
+        end
       end
     end
+
     # Allow this task to be called as many times as required. An example is the
     # migrate:redo task, which calls other two internally that depend on this one.
     db_namespace["_dump"].reenable
   end
 
   namespace :_dump do
-    ActiveRecord::Tasks::DatabaseTasks.for_each(databases) do |name|
+    ActiveRecord::Tasks::DatabaseTasks.for_each(databases) do |name, db_config|
       # IMPORTANT: This task won't dump the schema if ActiveRecord::Base.dump_schema_after_migration is set to false
       task name do
-        if ActiveRecord::Base.dump_schema_after_migration
```

Comment:
Once you remove returning `db_config` from `for_each` you need to do the following inside `task name`:

```ruby
db_config = ActiveRecord::Base.configurations.configs_for(env_name: ActiveRecord::Tasks::DatabaseTasks.env, name: name)
```